File: src/main/java/hashmap/GoodPairs.java
package hashmap;

import java.util.HashMap;

/**
 * https://leetcode-cn.com/problems/number-of-good-pairs/
 */
public class GoodPairs {
    public static void main(String args[]){
        int[] nums = {1,1,1,1};
        int pairs = new GoodPairs().numIdenticalPairs(nums);
        System.out.println(pairs);
    }

    private HashMap<Integer, Integer> dupNums;

    public int numIdenticalPairs(int[] nums) {
        dupNums = new HashMap<>(nums.length);
        for (int num : nums) {
            dupNums.put(num, dupNums.getOrDefault(num, 0) + 1);
        }

        int res = 0;
        for (Integer dupNum : dupNums.values()) {
            res += goodNum(dupNum);
        }
        return res;
    }

    private int goodNum(Integer dupNum) {
        if (dupNum < 2)
            return 0;
        if (dupNum == 2)
            return 1;
        return goodNum(dupNum-1) + (dupNum - 1);
    }
}

File: CppAlgorithms/hashmap/GoodPairs.cpp
#include <vector>
#include <unordered_map>
#include <iostream>
using namespace std;

/**
 * https://leetcode-cn.com/problems/number-of-good-pairs/
 *
 * 给你一个整数数组 nums 。
 * 如果一组数字 (i,j) 满足 nums[i] == nums[j] 且 i < j ，就可以认为这是一组 好数对 。
 * 返回好数对的数目。
 *
 * 输入：nums = [1,2,3,1,1,3]
 * 输出：4
 * 解释：有 4 组好数对，分别是 (0,3), (0,4), (3,4), (2,5) ，下标从 0 开始
 */
class Solution {
public:
    int numIdenticalPairs(vector<int>& nums) {
        unordered_map<int, int> num_cnt_map;
        int res = 0;
        for (int num: nums){
            res += num_cnt_map[num];
            ++num_cnt_map[num];
        }
        return res;
    }
};

int main() {
    vector<int> nums{1, 2, 3, 1, 1, 3};
    Solution s;
    cout << s.numIdenticalPairs(nums) << endl;
    return 0;
}